<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Обзор-данных-books" data-toc-modified-id="Обзор-данных-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обзор данных <code>books</code></a></span></li><li><span><a href="#Обзор-данных-authors" data-toc-modified-id="Обзор-данных-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обзор данных <code>authors</code></a></span></li><li><span><a href="#Обзор-данных-publishers" data-toc-modified-id="Обзор-данных-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обзор данных <code>publishers</code></a></span></li><li><span><a href="#Обзор-данных-ratings" data-toc-modified-id="Обзор-данных-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Обзор данных <code>ratings</code></a></span></li><li><span><a href="#Обзор-данных-reviews" data-toc-modified-id="Обзор-данных-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обзор данных <code>reviews</code></a></span></li></ul></li><li><span><a href="#Задания" data-toc-modified-id="Задания-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Посчитать,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитать,-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитать, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитать количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(так-мы-исключите-из-анализа-брошюры)" data-toc-modified-id="Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(так-мы-исключите-из-анализа-брошюры)-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определить издательство, которое выпустило наибольшее число книг толще 50 страниц (так мы исключите из анализа брошюры)</a></span></li><li><span><a href="#Определить-автора-с-самой-высокой-средней-оценкой-книг---учитываются-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определить-автора-с-самой-высокой-средней-оценкой-книг---учитываются-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определить автора с самой высокой средней оценкой книг - учитываются только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Задачи по SQL

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` - идентификатор книги;
- `author_id` - идентификатор автора;
- `title` - название книги;
- `num_pages` - количество страниц;
- `publication_date` - дата публикации книги;
- `publisher_id` - идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` - идентификатор автора;
- `author` - имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` - идентификатор издательства;
- `publisher` - название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` - идентификатор оценки;
- `book_id` - идентификатор книги;
- `username` - имя пользователя, оставившего оценку;
- `rating` - оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` - идентификатор обзора;
- `book_id` - идентификатор книги;
- `username` - имя пользователя, написавшего обзор;
- `text` - текст обзора.

In [1]:
from IPython.display import Image
from IPython.core.display import HTML
Image(url="https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?table=block&id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=2000&userId=&cache=v2", width=822, height=522)

## Обзор данных

In [2]:
!pip install sqlalchemy

In [3]:
# импортируем библиотеки

import pandas as pd
from sqlalchemy import create_engine

In [4]:
# устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

# сохраняем коннектор

con = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [5]:
# функция для первичного просмотра датасета

def data_look(data):
    # приводим названия всех колонок к змеиному регистру
    data.columns = [x.lower().replace(' ', '_') for x in data.columns]
    print()
    print('\033[1m === Информация по датасету === \033[0m')
    print()
    data.info() # выводим информацию по датасету
    print()
    print('\033[1m === Первые 5 строк данных === \033[0m')
    display(data.head()) # выводим первые 5 строк
    print()
    print('\033[1m === Случайные 5 строк данных === \033[0m')
    display(data.sample(5, random_state=10)) # выводим последние 5 строк  
    print()
    print('\033[1m === Пропуски и уникальные значения по столбцам === \033[0m')
    # выводим данные о пропусках и уникальных значениях по каждому столбцу
    display(pd.concat([data.isna().sum().rename('nan_count'), # количество пропусков
                       data.isna().mean().rename('nan_percent'), # % пропусков
                       data.nunique().rename('unique_values')], # количество уникальных значений
                      axis=1).style.format({'nan_percent' : '{:.2%}'})) 
    print()
    print('\033[1m === Явные дубликаты === \033[0m')
    print()
    print('Количество явных дубликатов:', data.duplicated().sum())

In [6]:
# функция для вывода результата по полученому SQL запросу

def select(sql):
    return pd.io.sql.read_sql(sql, con)

### Обзор данных `books`

In [7]:
# выполним запрос к базе данных и сохраним необходимую таблицу в переменную `books`

query_books = '''
          SELECT *
          FROM books
          LIMIT(5)
              '''
books = select(query_books)

In [8]:
data_look(books)


 === Информация по датасету === 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           5 non-null      int64 
 1   author_id         5 non-null      int64 
 2   title             5 non-null      object
 3   num_pages         5 non-null      int64 
 4   publication_date  5 non-null      object
 5   publisher_id      5 non-null      int64 
dtypes: int64(4), object(2)
memory usage: 368.0+ bytes

 === Первые 5 строк данных === 


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


 === Случайные 5 строк данных === 


book_id  author_id                                              title  \
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
0        1        546                                       'Salem's Lot   
4        5        125                                               1776   
1        2        465                 1 000 Places to See Before You Die   

   num_pages publication_date  publisher_id  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
0        594       2005-11-01            93  
4        386       2006-07-04           268  
1        992       2003-05-22           336


 === Пропуски и уникальные значения по столбцам === 



 === Явные дубликаты === 

Количество явных дубликатов: 0


В таблице `books` (содержит данные о книгах) 6 столбцов:

* `book_id` - идентификатор книги;
* `author_id` - идентификатор автора;
* `title` - название книги;
* `num_pages` - количество страниц;
* `publication_date` - дата публикации книги;
* `publisher_id` - идентификатор издателя.

Всего в таблице 1000 строк. Пропущенных значений и дубликатов нет.

### Обзор данных `authors`

In [9]:
# выполним запрос к базе данных и сохраним необходимую таблицу в переменную `authors`

query_authors = '''
            SELECT *
            FROM authors
                '''
authors = select(query_authors)

In [10]:
data_look(authors)


 === Информация по датасету === 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB

 === Первые 5 строк данных === 


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


 === Случайные 5 строк данных === 


author_id          author
27          28     Anna Sewell
475        476    Peter Straub
309        310   John Sandford
238        239       J.R. Ward
571        572  Tess Gerritsen


 === Пропуски и уникальные значения по столбцам === 



 === Явные дубликаты === 

Количество явных дубликатов: 0


В таблице `authors` (содержит данные об авторах) 2 столбца:

* `author_id` - идентификатор автора;
* `author` - имя автора.

Всего в таблице 636 строк. Пропущенных значений и дубликатов нет.

### Обзор данных `publishers`

In [11]:
# выполним запрос к базе данных и сохраним необходимую таблицу в переменную `publishers`

query_publishers = '''
               SELECT *
               FROM publishers
                   '''
publishers = select(query_publishers)

In [12]:
data_look(publishers)


 === Информация по датасету === 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB

 === Первые 5 строк данных === 


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


 === Случайные 5 строк данных === 


publisher_id                    publisher
322           323  W.W. Norton & Company (NYC)
227           228                       Pocket
323           324                 Walker Books
113           114                     Gollancz
80             81             Dell Laurel-Leaf


 === Пропуски и уникальные значения по столбцам === 



 === Явные дубликаты === 

Количество явных дубликатов: 0


В таблице `publishers` (содержит данные об издательствах) 2 столбца:

* `publisher_id` - идентификатор издательства;
* `publisher` - название издательства.

Всего в таблице 340 строк. Пропущенных значений и дубликатов нет.

### Обзор данных `ratings`

In [13]:
# выполним запрос к базе данных и сохраним необходимую таблицу в переменную `ratings`

query_ratings = '''
            SELECT *
            FROM ratings
                '''
ratings = select(query_ratings)

In [14]:
data_look(ratings)


 === Информация по датасету === 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB

 === Первые 5 строк данных === 


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


 === Случайные 5 строк данных === 


rating_id  book_id        username  rating
1901       1902      302  jennifermiller       3
330         331       67    abigailmoore       4
3559       3560      577      stacyfrost       4
3601       3602      583    ashleyjoshua       5
5044       5045      779   johnsonamanda       5


 === Пропуски и уникальные значения по столбцам === 



 === Явные дубликаты === 

Количество явных дубликатов: 0


В таблице `ratings` (содержит данные о пользовательских оценках книг) 4 столбца:

* `rating_id` - идентификатор оценки;
* `book_id` - идентификатор книги;
* `username` - имя пользователя, оставившего оценку;
* `rating` - оценка книги.

Всего в таблице 6456 строк. Пропущенных значений и дубликатов нет.

### Обзор данных `reviews`

In [15]:
# выполним запрос к базе данных и сохраним необходимую таблицу в переменную `reviews`

query_reviews = '''
            SELECT *
            FROM reviews
                '''
reviews = select(query_reviews)

In [16]:
data_look(reviews)


 === Информация по датасету === 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB

 === Первые 5 строк данных === 


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


 === Случайные 5 строк данных === 


review_id  book_id       username  \
311         310      108      emcdaniel   
1936       1934      698   martinyvonne   
1493       1492      542      bobbyhood   
862         863      317     martinadam   
774         773      286  taylorjeffrey   

                                                   text  
311   Although here season weight writer activity. S...  
1936  Recently director affect identify various sing...  
1493  Early culture control cut strategy manager wee...  
862   Month set where several. Game adult language o...  
774   Environment finally seek return water. All sou...


 === Пропуски и уникальные значения по столбцам === 



 === Явные дубликаты === 

Количество явных дубликатов: 0


В таблице `reviews` (содержит данные о пользовательских обзорах на книги) 4 столбца:

* `review_id` - идентификатор обзора;
* `book_id` - идентификатор книги;
* `username` - имя пользователя, написавшего обзор;
* `text` - текст обзора.

Всего в таблице 2793 строк. Пропущенных значений и дубликатов нет.

## Задания

### Посчитать, сколько книг вышло после 1 января 2000 года

In [17]:
query_1 = '''
    SELECT COUNT(book_id)
    FROM books
    WHERE publication_date > '2000-01-01';
        '''

df_1 = select(query_1)

print('C 1 января 2000 года выпущено {} книг'.format(df_1.loc[:, 'count'][0]))

C 1 января 2000 года выпущено 819 книг


### Для каждой книги посчитать количество обзоров и среднюю оценку

In [18]:
query_2 = '''
    WITH
    review_count AS
    
    (SELECT book_id,
            COUNT(review_id) AS cnt_review
     FROM reviews
     GROUP BY book_id),
     
    rating_agv AS
     
    (SELECT book_id,
            ROUND(AVG(rating), 2) AS avg_rating
     FROM ratings
     GROUP BY book_id)
      
    SELECT b.title,
           b.book_id,
           rc.cnt_review,
           ra.avg_rating
    FROM books b
    LEFT JOIN review_count rc ON b.book_id = rc.book_id
    LEFT JOIN rating_agv ra ON b.book_id = ra.book_id;
        '''
df_2 = select(query_2)
df_2

title  book_id  cnt_review  \
0            The Body in the Library (Miss Marple  #3)      652         2.0   
1                                            Galápagos      273         2.0   
2                             A Tree Grows in Brooklyn       51         5.0   
3    Undaunted Courage: The Pioneering First Missio...      951         2.0   
4                                          The Prophet      839         4.0   
..                                                 ...      ...         ...   
995   The Cat in the Hat and Other Dr. Seuss Favorites      672         NaN   
996    Anne Rice's The Vampire Lestat: A Graphic Novel       83         NaN   
997                          Essential Tales and Poems      221         NaN   
998                               Leonardo's Notebooks      387         NaN   
999                            The Natural Way to Draw      808         NaN   

     avg_rating  
0          4.50  
1          4.50  
2          4.25  
3          4.00  
4          4.29  
..          ...  
995        5.00  
996        3.67  
997        4.00  
998        4.00  
999        3.00  

[1000 rows x 4 columns]

### Определить издательство, которое выпустило наибольшее число книг толще 50 страниц (так мы исключите из анализа брошюры)

In [19]:
query_3 = '''
      SELECT b.publisher_id,
             p.publisher,
             COUNT(b.book_id) AS cnt_book
      FROM books b
      LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
      WHERE b.num_pages > 50
      GROUP BY b.publisher_id,
               p.publisher
      ORDER BY cnt_book DESC
      LIMIT(1);
           '''
df_3 = select(query_3)
print('Издательство {}, выпустило наибольшее число книг толще 50 страниц — {} шт.'.format(
    df_3.loc[:, 'publisher'][0], df_3.loc[:, 'cnt_book'][0]))

Издательство Penguin Books, выпустило наибольшее число книг толще 50 страниц — 42 шт.


### Определить автора с самой высокой средней оценкой книг - учитываются только книги с 50 и более оценками

In [20]:
query_4 = '''
    SELECT a.author,
           ROUND(AVG(r.rating), 2) AS avg_rating 
    FROM books b
    LEFT JOIN authors a ON b.author_id = a.author_id
    LEFT JOIN ratings r ON b.book_id = r.book_id 
    WHERE b.book_id IN (SELECT book_id
                        FROM ratings
                        GROUP BY book_id
                        HAVING COUNT(rating) >= 50)
    GROUP BY a.author
    ORDER BY avg_rating DESC
    LIMIT(1);
             '''
df_4 = select(query_4)
print('У автора {} самая высокая средняя оценка - {} среди книг с 50 и более оценками'.format(
    df_4.loc[:, 'author'][0], df_4.loc[:, 'avg_rating'][0]))

У автора J.K. Rowling/Mary GrandPré самая высокая средняя оценка - 4.29 среди книг с 50 и более оценками


### Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [21]:
query_5 = '''
    WITH
    count_review AS
    
    (SELECT username,
            COUNT(review_id)
    FROM reviews
    GROUP BY username)
    
    SELECT ROUND(AVG(count), 2)
    FROM count_review
    WHERE username IN (SELECT username
                       FROM ratings
                       GROUP BY username
                       HAVING COUNT(rating) > 50);
    '''
df_5 = select(query_5)
print('Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - {}'.format(df_5.loc[:, 'round'][0]))

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.33


## Вывод

1. C 1 января 2000 года выпущено 819 книг;
2. Издательство Penguin Books, выпустило наибольшее число книг толще 50 страниц — 42 шт.;
3. У автора J.K. Rowling/Mary GrandPré самая высокая средняя оценка - 4.28 среди книг с 50 и более оценками;
4. Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.33.